model/PermissionsService.php

```diff
@@ -54,61 +52,162 @@ public function __construct(
+     *   for that class and all its descendants (i.e. updates all resources AND
+     *   classes using the provided root class as the initial node for a tree
+     *   traversal, which may be slow and resource-intensive). This is provided
+     *   for backward compatibility purposes.
+     *
+     * - For non-recursive commands having a class as the root AND including
+     *   nested resources, updates permissions for the class and all instances
+     *   of that class, but skips all nested classes and instances of them (i.e.
+     *   does not go down into nested levels of the resource tree).
+     *
+     * - Otherwise (i.e. non-class roots), it updates only the resource set as
+     *   the root resource for the command.
+     */
+    public function applyPermissions(ChangePermissionsCommand $command): void
+    {
+        $resources = $this->getResourcesToUpdate($command);
+        $currentPermissions = $this->getResourcesPermissions($resources);
+        $permissionsDelta = $this->getDeltaForResourceTree($command, $currentPermissions);
 
-        if (empty($addRemove)) {
+        if (empty($permissionsDelta)) {
             return;
         }
 
-        $resourcesToUpdate = $this->getResourcesToUpdate($resource, $isRecursive);
-        $permissionsList = $this->getResourcesPermissions($resourcesToUpdate);
-        $actions = $this->getActions($resourcesToUpdate, $permissionsList, $addRemove);
-
-        $this->dryRun($actions, $permissionsList);
+        $actions = $this->getActions($resources, $currentPermissions, $permissionsDelta);
+        $this->dryRun($actions, $currentPermissions);
         $this->wetRun($actions);
-        $this->triggerEvents($addRemove, $resource->getUri(), $isRecursive);
+
+        $this->triggerEvents(
+            $permissionsDelta[$command->getRoot()->getUri()],
+            $command->getRoot()->getUri(),
+            $command->applyToNestedResources()
+        );
+    }
+
+    /**
+     * @deprecated Use applyPermissions() instead
+     * @see applyPermissions
+     */
+    public function saveResourcePermissionsRecursive(
+        core_kernel_classes_Resource $resource,
+        array $privilegesToSet
+    ): void {
+        $this->applyPermissions(
+            (new ChangePermissionsCommand($resource, $privilegesToSet))
+                ->withRecursion()
+                ->withNestedResources()
+        );
     }
 
+    /**
+     * @deprecated Use applyPermissions() instead
+     * @see applyPermissions
+     */
     public function savePermissions(
         bool $isRecursive,
         core_kernel_classes_Class $class,
         array $privilegesToSet
     ): void {
-        $this->saveResourcePermissions($class, $privilegesToSet, $isRecursive);
+        $this->applyPermissions(
+            (new ChangePermissionsCommand($class, $privilegesToSet))
+                ->withRecursion($isRecursive)
+                ->withNestedResources($isRecursive)
```

Comment:
I think we should not set both as true/false, but only the `isRecursive`. Wdyt?

```suggestion
            (new ChangePermissionsCommand($class, $privilegesToSet))
                ->withRecursion($isRecursive)
```

Reply:
This is needed to keep the old behavior: Previously, when `savePermissions()` was called to perform recursive calls, both nested resources (i.e. non-class instances) and classes themselves got their ACLs changed by this method if (and only if) `$isRecursive=true`.

Here we propagate the ACL changes to all classes and instances contained in the current class if `$isRecruisve=true`, otherwise we pass false to both parameters so only the root class is changed.

Reply:
Hmmm... actually, if `isRecursive()==true`, the new implementation uses `$resource-->getInstances(true)` to get all class and non-class instances under `$class` so, yes, setting just `isRecursive()` will keep the old behavior.

However, `triggerEvents()` will use `$command->applyToNestedResources()` to set `$isRecrusive` in `DataAccessControlChangedEvent`. I need to recheck since there may be a bug in how `triggerEvents` is called right now.

Reply:
Ok then, I wait your confirmation we are not producing any bug :) Thanks!

Reply:
`triggerEvents()` is now called using `$command->isRecursive()` and `$command->applyToNestedResources()` as its parameters, and the listener now discards events for classes if they both:

- Don't apply to nested resources (new behavior), and
- Don't apply recursively (old behavior)

I think we are fine.

Reply:
Ok, thanks for making sure ;)